TaskInputs.property(..., null) throws NPE

Plugins compiled against Gradle 4.3 throw an NPE if they try to do set an input property to `null` via Java code:

```java
getTaskInputs().property("some-property", null);
```

This produces the exception:

```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':foo'.
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:50)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:97)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:87)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:626)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:581)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.NullPointerException
        at org.gradle.api.internal.tasks.DefaultTaskInputPropertySpec.validate(DefaultTaskInputPropertySpec.java:53)
        at org.gradle.api.internal.tasks.DefaultTaskInputs.validate(DefaultTaskInputs.java:158)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:47)
        ... 24 more
```

The problem is caused by an internal overload introduced in 4.3:

```java
    TaskInputPropertyBuilder property(String name, ValidatingValue value);
```

Unfortunately, since `AbstractTask.getInputs()` returns the internal `TaskInputsInternal`, this method leaks into the public API.

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskInputPropertiesIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskInputPropertiesIntegrationTest.groovy
@@ -653,4 +653,40 @@ task someTask(type: SomeTask) {
         "dir"   | "input-file.txt" | true         | "Directory '<PATH>' specified for property 'output' is not a directory."
         "dirs"  | "input-file.txt" | true         | "Directory '<PATH>' specified for property 'output' is not a directory."
     }
+
+    def "can specify null as an input property in ad-hoc task"() {
+        buildFile << """
+            task foo {
+                inputs.property("a", null).optional(true)
+                doLast {}
+            }
+        """
+
+        expect:
+        succeeds "foo"
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/3366")
+    def "can specify null as an input property in Java task"() {
+        file("buildSrc/src/main/java/Foo.java") << """
+            import org.gradle.api.DefaultTask;
+            import org.gradle.api.tasks.TaskAction;
+
+            public class Foo extends DefaultTask {
+                public Foo() {
+                    getInputs().property("a", null).optional(true);
+                }
+                
+                @TaskAction
+                public void doSomething() {}
+            }
+        """
+
+        buildFile << """
+            task foo(type: Foo)
+        """
+
+        expect:
+        succeeds "foo"
+    }
 }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/TaskInputsInternal.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/TaskInputsInternal.java
@@ -27,15 +27,15 @@ import org.gradle.api.tasks.TaskInputs;
 public interface TaskInputsInternal extends TaskInputs {
     ImmutableSortedSet<TaskInputFilePropertySpec> getFileProperties();
 
-    TaskInputFilePropertyBuilderInternal files(ValidatingValue paths);
+    TaskInputFilePropertyBuilderInternal registerFiles(ValidatingValue paths);
 
-    TaskInputFilePropertyBuilderInternal file(ValidatingValue value);
+    TaskInputFilePropertyBuilderInternal registerFile(ValidatingValue value);
 
-    TaskInputFilePropertyBuilderInternal dir(ValidatingValue dirPath);
+    TaskInputFilePropertyBuilderInternal registerDir(ValidatingValue dirPath);
 
-    TaskInputPropertyBuilder property(String name, ValidatingValue value);
+    TaskInputPropertyBuilder registerProperty(String name, ValidatingValue value);
 
-    TaskInputPropertyBuilder nested(String name, ValidatingValue value);
+    TaskInputPropertyBuilder registerNested(String name, ValidatingValue value);
 
     void validate(TaskValidationContext context);
 }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/ClasspathPropertyAnnotationHandler.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/ClasspathPropertyAnnotationHandler.java
@@ -46,7 +46,7 @@ public class ClasspathPropertyAnnotationHandler implements OverridingPropertyAnn
     public void attachActions(final TaskPropertyActionContext context) {
         context.setConfigureAction(new UpdateAction() {
             public void update(TaskInternal task, TaskPropertyValue futureValue) {
-                task.getInputs().files(futureValue)
+                task.getInputs().registerFiles(futureValue)
                     .withPropertyName(context.getName())
                     .withNormalizer(ClasspathNormalizer.class)
                     .optional(context.isOptional());

--- a/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/CompileClasspathPropertyAnnotationHandler.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/CompileClasspathPropertyAnnotationHandler.java
@@ -46,7 +46,7 @@ public class CompileClasspathPropertyAnnotationHandler implements OverridingProp
     public void attachActions(final TaskPropertyActionContext context) {
         context.setConfigureAction(new UpdateAction() {
             public void update(TaskInternal task, TaskPropertyValue futureValue) {
-                task.getInputs().files(futureValue)
+                task.getInputs().registerFiles(futureValue)
                     .withPropertyName(context.getName())
                     .withNormalizer(CompileClasspathNormalizer.class)
                     .optional(context.isOptional());

--- a/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/InputDirectoryPropertyAnnotationHandler.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/InputDirectoryPropertyAnnotationHandler.java
@@ -28,6 +28,6 @@ public class InputDirectoryPropertyAnnotationHandler extends AbstractInputProper
     }
 
     protected TaskInputFilePropertyBuilder createPropertyBuilder(TaskPropertyActionContext context, TaskInternal task, TaskPropertyValue futureValue) {
-        return task.getInputs().dir(futureValue);
+        return task.getInputs().registerDir(futureValue);
     }
 }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/InputFilePropertyAnnotationHandler.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/InputFilePropertyAnnotationHandler.java
@@ -28,6 +28,6 @@ public class InputFilePropertyAnnotationHandler extends AbstractInputPropertyAnn
     }
 
     protected TaskInputFilePropertyBuilder createPropertyBuilder(TaskPropertyActionContext context, TaskInternal task, TaskPropertyValue futureValue) {
-        return task.getInputs().file(futureValue);
+        return task.getInputs().registerFile(futureValue);
     }
 }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/InputFilesPropertyAnnotationHandler.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/InputFilesPropertyAnnotationHandler.java
@@ -28,6 +28,6 @@ public class InputFilesPropertyAnnotationHandler extends AbstractInputPropertyAn
     }
 
     protected TaskInputFilePropertyBuilder createPropertyBuilder(TaskPropertyActionContext context, TaskInternal task, TaskPropertyValue futureValue) {
-        return task.getInputs().files(futureValue);
+        return task.getInputs().registerFiles(futureValue);
     }
 }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/InputPropertyAnnotationHandler.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/InputPropertyAnnotationHandler.java
@@ -32,7 +32,7 @@ public class InputPropertyAnnotationHandler implements PropertyAnnotationHandler
     public void attachActions(final TaskPropertyActionContext context) {
         context.setConfigureAction(new UpdateAction() {
             public void update(TaskInternal task, TaskPropertyValue futureValue) {
-                task.getInputs().property(context.getName(), futureValue)
+                task.getInputs().registerProperty(context.getName(), futureValue)
                     .optional(context.isOptional());
             }
         });

--- a/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/NestedBeanPropertyAnnotationHandler.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/NestedBeanPropertyAnnotationHandler.java
@@ -33,7 +33,7 @@ public class NestedBeanPropertyAnnotationHandler implements PropertyAnnotationHa
         context.setNestedType(context.getValueType());
         context.setConfigureAction(new UpdateAction() {
             public void update(TaskInternal task, final TaskPropertyValue futureValue) {
-                task.getInputs().nested(context.getName(), futureValue)
+                task.getInputs().registerNested(context.getName(), futureValue)
                     .optional(context.isOptional());
             }
         });

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskInputs.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskInputs.java
@@ -90,7 +90,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
         return taskMutator.mutate("TaskInputs.files(Object...)", new Callable<TaskInputFilePropertyBuilderInternal>() {
             @Override
             public TaskInputFilePropertyBuilderInternal call() {
-                return files(new StaticValue(unpackVarargs(paths)));
+                return registerFiles(new StaticValue(unpackVarargs(paths)));
             }
         });
     }
@@ -103,7 +103,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
     }
 
     @Override
-    public TaskInputFilePropertyBuilderInternal files(ValidatingValue paths) {
+    public TaskInputFilePropertyBuilderInternal registerFiles(ValidatingValue paths) {
         return addSpec(paths, ValidationAction.NO_OP);
     }
 
@@ -118,7 +118,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
     }
 
     @Override
-    public TaskInputFilePropertyBuilderInternal file(ValidatingValue value) {
+    public TaskInputFilePropertyBuilderInternal registerFile(ValidatingValue value) {
         return fileInternal(value, INPUT_FILE_VALIDATOR);
     }
 
@@ -137,7 +137,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
     }
 
     @Override
-    public TaskInputFilePropertyBuilderInternal dir(final ValidatingValue dirPath) {
+    public TaskInputFilePropertyBuilderInternal registerDir(final ValidatingValue dirPath) {
         return dir(dirPath, INPUT_DIRECTORY_VALIDATOR);
     }
 
@@ -217,7 +217,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
         return taskMutator.mutate("TaskInputs.property(String, Object)", new Callable<TaskInputPropertyBuilder>() {
             @Override
             public TaskInputPropertyBuilder call() {
-                return property(name, new StaticValue(value));
+                return registerProperty(name, new StaticValue(value));
             }
         });
     }
@@ -228,7 +228,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
             @Override
             public void run() {
                 for (Map.Entry<String, ?> entry : newProps.entrySet()) {
-                    property(entry.getKey(), new StaticValue(entry.getValue()));
+                    registerProperty(entry.getKey(), new StaticValue(entry.getValue()));
                 }
             }
         });
@@ -236,7 +236,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
     }
 
     @Override
-    public TaskInputPropertyBuilder property(String name, ValidatingValue value) {
+    public TaskInputPropertyBuilder registerProperty(String name, ValidatingValue value) {
         PropertyValue propertyValue = properties.get(name);
         DeclaredTaskInputProperty spec;
         if (propertyValue instanceof SimplePropertyValue) {
@@ -251,7 +251,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
     }
 
     @Override
-    public TaskInputPropertyBuilder nested(String name, ValidatingValue value) {
+    public TaskInputPropertyBuilder registerNested(String name, ValidatingValue value) {
         PropertyValue propertyValue = properties.get(name);
         DeclaredTaskInputProperty spec;
         if (propertyValue instanceof NestedBeanTypePropertyValue) {

